src/EventStore.Core/Caching/DynamicCacheManager.cs

```diff
@@ -0,0 +1,200 @@
+				stats[statNamePrefix + "mem-used"] = _cachesSettings[i].GetMemoryUsage();
+				stats[statNamePrefix + "mem-minAlloc"] = _cachesSettings[i].MinMemAllocation;
+				stats[statNamePrefix + "mem-maxAlloc"] = Interlocked.Read(ref _maxMemAllocation[i]);
+			}
+
+			message.Envelope.ReplyWith(new MonitoringMessage.InternalStatsRequestResponse(stats));
+		}
+
+		private void InitCacheSizes() {
+			var availableMem = CalcAvailableMemory(_getFreeMem(), 0L);
+			var cacheIndex = -1;
+			foreach (var cacheSettings in _cachesSettings) {
+				cacheIndex++;
+
+				if (!cacheSettings.IsDynamic) {
+					Log.Information("{name} cache size configured to ~{configuredMem:N0} bytes.",
+						cacheSettings.Name, cacheSettings.InitialMaxMemAllocation);
+					_maxMemAllocation[cacheIndex] = cacheSettings.InitialMaxMemAllocation;
+				} else {
+					var allocatedMem = CalcMemAllocation(availableMem, cacheSettings.Weight, cacheSettings.MinMemAllocation);
+					cacheSettings.InitialMaxMemAllocation = allocatedMem;
+					Log.Information(
+						"{name} cache size auto-configured to ~{allocatedMem:N0} bytes.",
+						cacheSettings.Name, allocatedMem);
+					_maxMemAllocation[cacheIndex] = allocatedMem;
+				}
+			}
+		}
+
+		private void ResizeCachesIfNeeded() {
+			var freeMem = _getFreeMem();
+			var keepFreeMem = Math.Max(_keepFreeMemBytes, _totalMem * _keepFreeMemPercent / 100);
+
+			if (freeMem >= keepFreeMem && DateTime.UtcNow - _lastResize < _minResizeInterval)
+				return;
+
+			if (freeMem < keepFreeMem) {
+				Log.Debug("Available system memory is lower than "
+				          + "{thresholdPercent}% or {thresholdBytes:N0} bytes: {freeMem:N0} bytes. Resizing caches.",
+					_keepFreeMemPercent, _keepFreeMemBytes, freeMem);
+			}
+
+			try {
+				var cachedMem = _cachesSettings.Sum(cacheSettings => cacheSettings.GetMemoryUsage());
+				ResizeCaches(freeMem, cachedMem);
+			} finally {
+				_lastResize = DateTime.UtcNow;
+			}
+		}
+
+		private void ResizeCaches(long freeMem, long cachedMem) {
+			var availableMem = CalcAvailableMemory(freeMem, cachedMem);
+
+			var sw = new Stopwatch();
+			var cacheIndex = -1;
+
+			foreach (var cacheSettings in _cachesSettings) {
+				cacheIndex++;
+
+				if (!cacheSettings.IsDynamic)
+					continue;
+
+				var allocatedMem = CalcMemAllocation(availableMem, cacheSettings.Weight, cacheSettings.MinMemAllocation);
+
+				// do not resize if the amount of memory allocated to the cache hasn't changed
+				if (_maxMemAllocation[cacheIndex] == allocatedMem)
+					continue;
+
+				sw.Restart();
+				cacheSettings.UpdateMaxMemoryAllocation(allocatedMem);
+				sw.Stop();
+				Log.Debug(
+					"{name} cache resized to ~{allocatedMem:N0} bytes in {elapsed}.",
+					cacheSettings.Name, allocatedMem, sw.Elapsed);
+				_maxMemAllocation[cacheIndex] = allocatedMem;
+			}
+		}
+
+		private long CalcAvailableMemory(long freeMem, long cachedMem) {
+			var keepFreeMem = Math.Max(_keepFreeMemBytes, _totalMem * _keepFreeMemPercent / 100);
```

Comment:
dunno when machines will get enough memory to make this overflow but :)

Reply:
fixed in: c2d1aac9ea0cc79ff2f0ae338223c45a936a4e82